README.md
# kubernetes-ssl-manager

# Problem
When deploying services to Kubernetes, a certificate has to be injected into the container via secret. It doesn't make sense to have each container renew it's own certificates as it's state can be wiped at any given time.

# Solution
Build a service within each Kubernetes namespace that handles renewing all certificates used in that namespace. This service would kick off the request to renew each cert at a predetermined interval. It would then accept all verification requests ( GET request to domain/.well-known/acme-challenge ) and respond as necessary. After being issued the new certificate, it would recreate the appropriate secret which contains that certificate and initiate a restart of any container or service necessary.

Available on docker hub as [phutchins/kubernetes-ssl-manager](https://hub.docker.com/r/phutchins/kubernetes-ssl-manager)

## Useful commands

### Generate a new set of certs

Once this container is running you can generate new certificates using:

```
kubectl exec -it <pod> -- bash -c 'EMAIL=fred@fred.com DOMAINS=example.com foo.example.com ./fetch_certs.sh'
```

### Save the set of certificates as a secret

```
kubectl exec -it <pod> -- bash -c 'DOMAINS=example.com foo.example.com ./save_certs.sh'
```

### Refresh the certificates

```
kubectl exec -it <pod> -- bash -c 'EMAIL=fred@fred.com DOMAINS=example.com foo.example.com SECRET_NAME=foo DEPLOYMENTS=bar ./refresh_certs.sh'
```

## Environment variables:

 - EMAIL - the email address to obtain certificates on behalf of.
 - DOMAINS - a space separated list of domains to obtain a certificate for.
 - LETSENCRYPT_ENDPOINT
   - If set, will be used to populate the /etc/letsencrypt/cli.ini file with
     the given server value. For testing use
     https://acme-staging.api.letsencrypt.org/directory
 - DEPLOYMENTS - a space separated list of deployments whose pods should be
   refreshed after a certificate save
 - SECRET_NAME - the name to save the secrets under
 - NAMESPACE - the namespace under which the secrets should be available
 - CRON_FREQUENCY - the 5-part frequency of the cron job. Default is a random
   time in the range `0-59 0-23 1-27 * *`


CHANGELOG.md
# 0.0.9

 - NAMESPACE bugfixes
 
# 0.0.8

 - NAMESPACE env var support added

# 0.0.7

 - Upgrade kubectl so that deployments work

# 0.0.6

 - Added support for k8s deployments
 - Removed support for RC based refreshes

# 0.0.5


save_certs.sh
#!/bin/bash

# $DOMAINS should contain all domains that this container is responsible for
# renewing. The first one dictates where the cert will live.

# Inside /etc/letsencrypt/live/<domain> we have:
#
# cert.pem  chain.pem  fullchain.pem  privkey.pem
#
# We want to convert fullchain.pem into proxycert
# and privkey.pem into proxykey and then save as a secret!

if [ -z "$SECRET_PREFIX" ]; then
  SECRET_PREFIX=cert
fi


CERT_LOCATION='/etc/letsencrypt/live'

MAIN_DOMAIN=${DOMAINS[0]}

for DOMAIN in $DOMAINS; do
  SECRET_NAME="$SECRET_PREFIX.$DOMAIN"

  CERT=$(cat $CERT_LOCATION/$MAIN_DOMAIN/fullchain.pem | base64 --wrap=0)
  KEY=$(cat $CERT_LOCATION/$MAIN_DOMAIN/privkey.pem | base64 --wrap=0)
  DHPARAM=$(openssl dhparam 2048 | base64 --wrap=0)

  NAMESPACE=${NAMESPACE:-default}

  EXPIRE_DATE=$(openssl x509 -enddate -noout -in $CERT_LOCATION/$MAIN_DOMAIN/fullchain.pem | awk -F "=" '{print $2}' | base64 --wrap=0)

  kubectl get secrets --namespace $NAMESPACE $SECRET_NAME && ACTION=replace || ACTION=create;

  cat << EOF | kubectl $ACTION -f -
  {
   "apiVersion": "v1",
   "kind": "Secret",
   "metadata": {
     "labels": {
       "type": "cert"
     },
     "name": "$SECRET_NAME",
     "namespace": "$NAMESPACE"
   },
   "data": {
     "tls.crt": "$CERT",
     "tls.key": "$KEY",
     "tls.dhparam": "$DHPARAM",
     "tls.expires": "$EXPIRE_DATE"
   }
  }
EOF

done


recreate_pods.sh
#!/bin/bash

# Update the required env vars for the first pod in each Deployment.
# This will kick off a rolling update.
# Do this so that the secrets can be remounted.

if [ -z "$DEPLOYMENTS" ]; then
    echo "WARNING: DEPLOYMENTS not provided. Secret changes may not be reflected."
    exit
fi

DEPLOYMENTS=(${DEPLOYMENTS})
DATE=$(date)
NAMESPACE=${NAMESPACE:-default}

for DEPLOYMENT in $DEPLOYMENTS
do
    NAME=$(kubectl get deployments --namespace $NAMESPACE $DEPLOYMENT -o=template --template='{{index .spec.template.spec.containers 0 "name"}}')
    PATCH=$(NAME=$NAME DATE=$DATE echo "{\"spec\": {\"template\": {\"spec\": {\"containers\": [{\"name\": \"$NAME\", \"env\": [{\"name\": \"LETSENCRYPT_CERT_REFRESH\", \"value\": \"$DATE\"}]}]}}}}")
    echo "PATCHING ${DEPLOYMENT}: ${PATCH}"
    kubectl patch deployment --namespace $NAMESPACE $DEPLOYMENT --type=strategic -p "$PATCH"
done


start.sh
#!/bin/bash

# Add a cron line with details of the current user etc
minute=$(echo $RANDOM % 60 | bc)
hour=$(echo $RANDOM % 23 | bc)
day=$(echo $RANDOM % 27 + 1 | bc)

CRON_FREQUENCY=${CRON_FREQUENCY:-"$minute $hour $day * *"}
NAMESPACE=${NAMESPACE:-default}

echo "Configuring cron..."
echo "DOMAINS: " $DOMAINS
echo "EMAIL: " $EMAIL
echo "DEPLOYMENTS: " $DEPLOYMENTS
echo "NAMESPACE: " $NAMESPACE
echo "SECRET_NAME: " $SECRET_NAME
echo "CRON frequency: " $CRON_FREQUENCY
# Once a month, fetch and save certs + restart pods.

# The process running under cron needs to know where the to find the kubernetes api
env_vars="PATH=$PATH KUBERNETES_PORT=$KUBERNETES_PORT KUBERNETES_PORT_443_TCP_PORT=$KUBERNETES_PORT_443_TCP_PORT KUBERNETES_SERVICE_PORT=$KUBERNETES_SERVICE_PORT KUBERNETES_SERVICE_HOST=$KUBERNETES_SERVICE_HOST KUBERNETES_PORT_443_TCP_PROTO=$KUBERNETES_PORT_443_TCP_PROTO KUBERNETES_PORT_443_TCP_ADDR=$KUBERNETES_PORT_443_TCP_ADDR KUBERNETES_PORT_443_TCP=$KUBERNETES_PORT_443_TCP"

line="$CRON_FREQUENCY $env_vars SECRET_NAME=$SECRET_NAME NAMESPACE=$NAMESPACE DEPLOYMENTS='$DEPLOYMENTS' DOMAINS='$DOMAINS' EMAIL=$EMAIL /bin/bash /letsencrypt/refresh_certs.sh >> /var/log/cron-encrypt.log 2>&1"
(crontab -u root -l; echo "$line" ) | crontab -u root -

if [ -n "${LETSENCRYPT_ENDPOINT+1}" ]; then
    echo "server = $LETSENCRYPT_ENDPOINT" >> /etc/letsencrypt/cli.ini
fi

# Start cron
echo "Starting cron..."
cron &

echo "Starting nginx..."
nginx -g 'daemon off;'
